### PR TITLE
Don't publish private entries

### PR DESCRIPTION
**Pull request by struktured**
Please check one of the following, relating to the [CHANGELOG](https://github.com/holochain/holochain-rust/blob/develop/CHANGELOG.md), which should be updated if relevant
- [x] my changes to the code affect some exposed aspect of the developer experience, and I have added an item to relevant 'Added, Fixed, Changed, Removed, Deprecated, or Security' heading under the 'Unreleased' heading of the CHANGELOG, with the format `- summary of change [PR#1234](https://github.com/holochain/holochain-rust/pull/1234)`
- [ ] my changes to the code do not affect any exposed aspect of the developer experience

This introduces a fix so that entries defined as being private are not published to the DHT. Currently entry types do not examine there own definitions to ascertain their sharing attribute- this work enhances the publishable logic accordingly.  

As this is my first project with the holochain core team, it is requested that reviewers are particularly thorough and stringent before accepting the PR.

# Extend EntryType with CanPublish trait

 - [x] Remove `can_publish` function from existing entry type definition in the `core_types` crate.
 - [x] Introduce `CanPublish` trait in `core` crate which requires a `Context` to properly determine whether the entry type is sharable.
 - [x] Have `EntryType` implement new `CanPublish` trait by extending the entry type functionality in the `core` crate (even though the `EntryType` itself is still defined `core_types`). Most of [this commented out code](https://github.com/holochain/holochain-rust/blob/develop/core_types/src/entry/entry_type.rs#L96) by @lucksus can be probably reused for this purpose.

# Rust tests

 - [x] Add test module in `core` which creates a just complicated enough `Context` to verify private entries are _not_ published to the DHT. (consider using / extended the existing holochain test library crate).

# App-spec tests
 functions
 - [x] Make a private entry type called `memo` to the `blog` zome which represents private notes and reminders users only write to their local chain. Add public zome functions `hash_memo`, `create_memo`, and `my_memos`.
 - [x] Test that the new entry type is creatable and queryable with aforementioned zome functions.
 - [x] Ensure other users cannot find these entries in their own local chains or the DHT.
 - [ ] Test that the chain header still contains the entry address and relevant metadata but _without_ the entry itself. **N/A:** Considered out of scope for this project

### CODE REVIEW

**Comment by thedavidmeister** [comment]
mmm seems like something we want

File: doc/holochain_101/src/faq.md [hunk 1]
```diff
@@ -65,7 +65,7 @@ DHTs enable key/value pair storage and retrieval across many machines. The only
 
 In fact, since many DHTs are used for illegal file sharing (Napster, Bittorrent, Sharezaa, etc.), they are designed to protect anonymity of uploaders so they won't get in trouble. File sharing DHTs frequently serve virus infected files, planted by uploaders trying to infect digital pirates. There's no accountability for actions or reliable way to ensure bad data doesn't spread.
```
Comment (zippy) [inline doc/holochain_101/src/faq.md]: This file also shouldn't show up as changed in this PR right?

File: cas_implementations/src/cas/pickle.rs [hunk 2]
```diff
@@ -10,9 +10,12 @@ use std::{
     fmt::{Debug, Error, Formatter},
     path::Path,
     sync::{Arc, RwLock},
+    time::Duration,
```
[the remark this answers is not in the document]
Comment (struktured) [reply cas_implementations/src/cas/pickle.rs]: Not sure, either- possibly an awry rebase- will ensure it's clean.

File: app_spec/zomes/blog/code/src/memo.rs [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+                   .ok_or_else(|| String::from("Content too long"))
+                },
+                EntryValidationData::Modify{new_entry:new_memo,old_entry:old_memo,old_entry_header:_,validation_data:_} =>
+                {
+                   (new_memo.content != old_memo.content)
+                   .ok_or_else(|| String::from("Trying to modify with same data"))
+                },
+                EntryValidationData::Delete{old_entry:old_memo,old_entry_header:_,validation_data:_} =>
+                {
+                   (old_memo.content!="SYS")
```
[the remark this answers is not in the document]
Comment (Connoropolous) [reply app_spec/zomes/blog/code/src/memo.rs]: It's there cause it's copied from the other zome code of `blog` zome I believe. So if we question it here we should question it there

**Comment by struktured** [comment]
It should be removed- it's not really testing any useful functionality
here- I will remove it.

File: app_spec/zomes/blog/code/src/memo.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (struktured) [reply app_spec/zomes/blog/code/src/memo.rs]: See commit 8d04d680c456645714bbe9f21142dac807b810c4.

File: core/src/workflows/get_entry_result.rs [hunk 4]
```diff
@@ -21,11 +21,30 @@ pub async fn get_entry_with_meta_workflow<'a>(
...
+            address
+        ));
+        let entry_result = await!(network::actions::get_entry::get_entry(
             context.clone(),
             address.clone(),
             timeout.clone(),
-        ))
+        ));
+
+        match entry_result.clone() {
```
[the remark this answers is not in the document]
Comment (struktured) [reply core/src/workflows/get_entry_result.rs]: See commit 8dd9f9e8ebab9aad1f16cb9a70e08c993bdf0455.

[the remark this answers is not in the document]
Comment (struktured) [reply core/src/workflows/get_entry_result.rs]: @zippy you mentioned `handle_fetch_entry` _and_ `handle_fetch_entry_result`- are both required to check `can_publish`? I only added it for `handle_fetch_entry` and it worked with my tests at least. Please confirm desired behavior for `handle_fetch_entry_result`.

**Pull request merged**